File: src/features2d/features2d_async.cpp
/* Created by Abdelaziz Mahdy. Licensed: Apache 2.0 license. Copyright (c) 2024 Abdelaziz Mahdy. */
#include "features2d_async.h"
#include "utils.hpp"

#include "core/types.h"
#include "core/vec.hpp"

// Asynchronous functions for AKAZE
CvStatus *AKAZE_Create_Async(CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new AKAZE{new cv::Ptr<cv::AKAZE>(cv::AKAZE::create())});
  END_WRAP
}

CvStatus *AKAZE_Close_Async(AKAZEPtr self, CvCallback_0 callback) {
  BEGIN_WRAP
  self->ptr->reset();
  CVD_FREE(self);
  callback();
  END_WRAP
}

CvStatus *AKAZE_Detect_Async(AKAZE self, Mat src, CvCallback_1 callback) {
  BEGIN_WRAP
  std::vector<cv::KeyPoint> detected;
  (*self.ptr)->detect(*src.ptr, detected);
  callback(veckeypoint_cpp2c_p(detected));
  END_WRAP
}

CvStatus *AKAZE_DetectAndCompute_Async(AKAZE self, Mat src, Mat mask, CvCallback_2 callback) {
  BEGIN_WRAP
  std::vector<cv::KeyPoint> detected;
  cv::Mat desc;
  (*self.ptr)->detectAndCompute(*src.ptr, *mask.ptr, detected, desc);
  callback(veckeypoint_cpp2c_p(detected), new Mat{new cv::Mat(desc)});
  END_WRAP
}

// Asynchronous functions for AgastFeatureDetector
CvStatus *AgastFeatureDetector_Create_Async(CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new AgastFeatureDetector{
      new cv::Ptr<cv::AgastFeatureDetector>(cv::AgastFeatureDetector::create())
  });
  END_WRAP
}

CvStatus *AgastFeatureDetector_Close_Async(AgastFeatureDetectorPtr self, CvCallback_0 callback) {
  BEGIN_WRAP
  self->ptr->reset();
  CVD_FREE(self);
  callback();
  END_WRAP
}

CvStatus *
AgastFeatureDetector_Detect_Async(AgastFeatureDetector self, Mat src, CvCallback_1 callback) {
  BEGIN_WRAP
  std::vector<cv::KeyPoint> detected;
  (*self.ptr)->detect(*src.ptr, detected);
  callback(veckeypoint_cpp2c_p(detected));
  END_WRAP
}

// Asynchronous functions for BRISK
CvStatus *BRISK_Create_Async(CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new BRISK{new cv::Ptr<cv::BRISK>(cv::BRISK::create())});
  END_WRAP
}

CvStatus *BRISK_Close_Async(BRISKPtr self, CvCallback_0 callback) {
  BEGIN_WRAP
  self->ptr->reset();
  CVD_FREE(self);
  callback();
  END_WRAP
}

CvStatus *BRISK_Detect_Async(BRISK self, Mat src, CvCallback_1 callback) {
  BEGIN_WRAP
  std::vector<cv::KeyPoint> detected;
  (*self.ptr)->detect(*src.ptr, detected);
  callback(veckeypoint_cpp2c_p(detected));
  END_WRAP
}

CvStatus *BRISK_DetectAndCompute_Async(BRISK self, Mat src, Mat mask, CvCallback_2 callback) {
  BEGIN_WRAP
  std::vector<cv::KeyPoint> detected;
  cv::Mat desc;
  (*self.ptr)->detectAndCompute(*src.ptr, *mask.ptr, detected, desc);
  callback(veckeypoint_cpp2c_p(detected), new Mat{new cv::Mat(desc)});
  END_WRAP
}

// Asynchronous functions for FastFeatureDetector
CvStatus *FastFeatureDetector_Create_Async(CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new FastFeatureDetector{
      new cv::Ptr<cv::FastFeatureDetector>(cv::FastFeatureDetector::create())
  });
  END_WRAP
}

CvStatus *FastFeatureDetector_CreateWithParams_Async(
    int threshold, bool nonmaxSuppression, int type, CvCallback_1 callback
) {
  BEGIN_WRAP
  auto type_ = static_cast<cv::FastFeatureDetector::DetectorType>(type);
  callback(new FastFeatureDetector{new cv::Ptr<cv::FastFeatureDetector>(
      cv::FastFeatureDetector::create(threshold, nonmaxSuppression, type_)
  )});
  END_WRAP
}

CvStatus *FastFeatureDetector_Close_Async(FastFeatureDetectorPtr self, CvCallback_0 callback) {
  BEGIN_WRAP
  self->ptr->reset();
  CVD_FREE(self);
  callback();
  END_WRAP
}

CvStatus *
FastFeatureDetector_Detect_Async(FastFeatureDetector self, Mat src, CvCallback_1 callback) {
  BEGIN_WRAP
  std::vector<cv::KeyPoint> detected;
  (*self.ptr)->detect(*src.ptr, detected);
  callback(veckeypoint_cpp2c_p(detected));
  END_WRAP
}

// Asynchronous functions for GFTTDetector
CvStatus *GFTTDetector_Create_Async(CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new GFTTDetector{new cv::Ptr<cv::GFTTDetector>(cv::GFTTDetector::create())});
  END_WRAP
}

CvStatus *GFTTDetector_Close_Async(GFTTDetectorPtr self, CvCallback_0 callback) {
  BEGIN_WRAP
  self->ptr->reset();
  CVD_FREE(self);
  callback();
  END_WRAP
}

CvStatus *GFTTDetector_Detect_Async(GFTTDetector self, Mat src, CvCallback_1 callback) {
  BEGIN_WRAP
  std::vector<cv::KeyPoint> detected;
  (*self.ptr)->detect(*src.ptr, detected);
  callback(veckeypoint_cpp2c_p(detected));
  END_WRAP
}

// Asynchronous functions for KAZE
CvStatus *KAZE_Create_Async(CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new KAZE{new cv::Ptr<cv::KAZE>(cv::KAZE::create())});
  END_WRAP
}

CvStatus *KAZE_Close_Async(KAZEPtr self, CvCallback_0 callback) {
  BEGIN_WRAP
  self->ptr->reset();
  CVD_FREE(self);
  callback();
  END_WRAP
}

CvStatus *KAZE_Detect_Async(KAZE self, Mat src, CvCallback_1 callback) {
  BEGIN_WRAP
  std::vector<cv::KeyPoint> detected;
  (*self.ptr)->detect(*src.ptr, detected);
  callback(veckeypoint_cpp2c_p(detected));
  END_WRAP
}

CvStatus *KAZE_DetectAndCompute_Async(KAZE self, Mat src, Mat mask, CvCallback_2 callback) {
  BEGIN_WRAP
  std::vector<cv::KeyPoint> detected;
  cv::Mat desc;
  (*self.ptr)->detectAndCompute(*src.ptr, *mask.ptr, detected, desc);
  callback(veckeypoint_cpp2c_p(detected), new Mat{new cv::Mat(desc)});
  END_WRAP
}

// Asynchronous functions for MSER
CvStatus *MSER_Create_Async(CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new MSER{new cv::Ptr<cv::MSER>(cv::MSER::create())});
  END_WRAP
}

CvStatus *MSER_Close_Async(MSERPtr self, CvCallback_0 callback) {
  BEGIN_WRAP
  self->ptr->reset();
  CVD_FREE(self);
  callback();
  END_WRAP
}

CvStatus *MSER_Detect_Async(MSER self, Mat src, CvCallback_1 callback) {
  BEGIN_WRAP
  std::vector<cv::KeyPoint> detected;
  (*self.ptr)->detect(*src.ptr, detected);
  callback(veckeypoint_cpp2c_p(detected));
  END_WRAP
}

// Asynchronous functions for ORB
CvStatus *ORB_Create_Async(CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new ORB{new cv::Ptr<cv::ORB>(cv::ORB::create())});
  END_WRAP
}

CvStatus *ORB_CreateWithParams_Async(
    int nfeatures,
    float scaleFactor,
    int nlevels,
    int edgeThreshold,
    int firstLevel,
    int WTA_K,
    int scoreType,
    int patchSize,
    int fastThreshold,
    CvCallback_1 callback
) {
  BEGIN_WRAP
  auto type = static_cast<cv::ORB::ScoreType>(scoreType);
  callback(new ORB{new cv::Ptr<cv::ORB>(cv::ORB::create(
      nfeatures,
      scaleFactor,
      nlevels,
      edgeThreshold,
      firstLevel,
      WTA_K,
      type,
      patchSize,
      fastThreshold
  ))});
  END_WRAP
}

CvStatus *ORB_Close_Async(ORBPtr self, CvCallback_0 callback) {
  BEGIN_WRAP
  self->ptr->reset();
  CVD_FREE(self);
  callback();
  END_WRAP
}

CvStatus *ORB_Detect_Async(ORB self, Mat src, CvCallback_1 callback) {
  BEGIN_WRAP
  std::vector<cv::KeyPoint> detected;
  (*self.ptr)->detect(*src.ptr, detected);
  callback(veckeypoint_cpp2c_p(detected));
  END_WRAP
}

CvStatus *ORB_DetectAndCompute_Async(ORB self, Mat src, Mat mask, CvCallback_2 callback) {
  BEGIN_WRAP
  std::vector<cv::KeyPoint> detected;
  cv::Mat desc;
  (*self.ptr)->detectAndCompute(*src.ptr, *mask.ptr, detected, desc);
  callback(veckeypoint_cpp2c_p(detected), new Mat{new cv::Mat(desc)});
  END_WRAP
}

// Asynchronous functions for SimpleBlobDetector
CvStatus *SimpleBlobDetector_Create_Async(CvCallback_1 callback) {
  BEGIN_WRAP
  callback(
      new SimpleBlobDetector{new cv::Ptr<cv::SimpleBlobDetector>(cv::SimpleBlobDetector::create())}
  );
  END_WRAP
}

CvStatus *
SimpleBlobDetector_Create_WithParams_Async(SimpleBlobDetectorParams params, CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new SimpleBlobDetector{new cv::Ptr<cv::SimpleBlobDetector>(
      cv::SimpleBlobDetector::create(ConvertCParamsToCPPParams(params))
  )});
  END_WRAP
}

CvStatus *SimpleBlobDetector_Close_Async(SimpleBlobDetectorPtr self, CvCallback_0 callback) {
  BEGIN_WRAP
  self->ptr->reset();
  CVD_FREE(self);
  callback();
  END_WRAP
}

CvStatus *SimpleBlobDetector_Detect_Async(SimpleBlobDetector self, Mat src, CvCallback_1 callback) {
  BEGIN_WRAP
  std::vector<cv::KeyPoint> detected;
  (*self.ptr)->detect(*src.ptr, detected);
  callback(veckeypoint_cpp2c_p(detected));
  END_WRAP
}

// Asynchronous functions for BFMatcher
CvStatus *BFMatcher_Create_Async(CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new BFMatcher{new cv::Ptr<cv::BFMatcher>(cv::BFMatcher::create())});
  END_WRAP
}

CvStatus *BFMatcher_CreateWithParams_Async(int normType, bool crossCheck, CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new BFMatcher{new cv::Ptr<cv::BFMatcher>(cv::BFMatcher::create(normType, crossCheck))});
  END_WRAP
}

CvStatus *BFMatcher_Close_Async(BFMatcherPtr self, CvCallback_0 callback) {
  BEGIN_WRAP
  self->ptr->reset();
  CVD_FREE(self);
  callback();
  END_WRAP
}

CvStatus *BFMatcher_Match_Async(BFMatcher self, Mat query, Mat train, CvCallback_1 callback) {
  BEGIN_WRAP
  std::vector<cv::DMatch> matches;
  (*self.ptr)->match(*query.ptr, *train.ptr, matches);
  callback(vecdmatch_cpp2c_p(matches));
  END_WRAP
}

CvStatus *
BFMatcher_KnnMatch_Async(BFMatcher self, Mat query, Mat train, int k, CvCallback_1 callback) {
  BEGIN_WRAP
  std::vector<std::vector<cv::DMatch>> matches;
  (*self.ptr)->knnMatch(*query.ptr, *train.ptr, matches, k);
  callback(vecvecdmatch_cpp2c_p(matches));
  END_WRAP
}

// Asynchronous functions for FlannBasedMatcher
CvStatus *FlannBasedMatcher_Create_Async(CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new FlannBasedMatcher{new cv::Ptr<cv::FlannBasedMatcher>(cv::FlannBasedMatcher::create())
  });
  END_WRAP
}

CvStatus *FlannBasedMatcher_Close_Async(FlannBasedMatcherPtr self, CvCallback_0 callback) {
  BEGIN_WRAP
  self->ptr->reset();
  CVD_FREE(self);
  callback();
  END_WRAP
}

CvStatus *FlannBasedMatcher_KnnMatch_Async(
    FlannBasedMatcher self, Mat query, Mat train, int k, CvCallback_1 callback
) {
  BEGIN_WRAP
  std::vector<std::vector<cv::DMatch>> matches;
  (*self.ptr)->knnMatch(*query.ptr, *train.ptr, matches, k);
  callback(vecvecdmatch_cpp2c_p(matches));
  END_WRAP
}

// Asynchronous utility functions
CvStatus *DrawKeyPoints_Async(
    Mat src, VecKeyPoint kp, Mat dst, const Scalar color, int flags, CvCallback_0 callback
) {
  BEGIN_WRAP
  auto color_ = cv::Scalar(color.val1, color.val2, color.val3, color.val4);
  auto _kp = veckeypoint_c2cpp(kp);
  cv::drawKeypoints(*src.ptr, _kp, *dst.ptr, color_, static_cast<cv::DrawMatchesFlags>(flags));
  callback();
  END_WRAP
}

CvStatus *DrawMatches_Async(
    Mat img1,
    VecKeyPoint kp1,
    Mat img2,
    VecKeyPoint kp2,
    VecDMatch matches1to2,
    Mat outImg,
    const Scalar matchesColor,
    const Scalar pointColor,
    VecChar matchesMask,
    int flags,
    CvCallback_0 callback
) {
  BEGIN_WRAP
  auto mColor =
      cv::Scalar(matchesColor.val1, matchesColor.val2, matchesColor.val3, matchesColor.val4);
  auto pColor = cv::Scalar(pointColor.val1, pointColor.val2, pointColor.val3, pointColor.val4);
  auto _kp1 = veckeypoint_c2cpp(kp1);
  auto _kp2 = veckeypoint_c2cpp(kp2);
  auto _matches1to2 = vecdmatch_c2cpp(matches1to2);
  auto _matchesMask = vecchar_c2cpp(matchesMask);
  cv::drawMatches(
      *img1.ptr,
      _kp1,
      *img2.ptr,
      _kp2,
      _matches1to2,
      *outImg.ptr,
      mColor,
      pColor,
      _matchesMask,
      static_cast<cv::DrawMatchesFlags>(flags)
  );
  callback();
  END_WRAP
}

// Asynchronous functions for SIFT
CvStatus *SIFT_Create_Async(CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new SIFT{new cv::Ptr<cv::SIFT>(cv::SIFT::create())});
  END_WRAP
}

CvStatus *SIFT_Close_Async(SIFTPtr self, CvCallback_0 callback) {
  BEGIN_WRAP
  self->ptr->reset();
  CVD_FREE(self);
  callback();
  END_WRAP
}

CvStatus *SIFT_Detect_Async(SIFT self, Mat src, CvCallback_1 callback) {
  BEGIN_WRAP
  std::vector<cv::KeyPoint> detected;
  (*self.ptr)->detect(*src.ptr, detected);
  callback(veckeypoint_cpp2c_p(detected));
  END_WRAP
}

CvStatus *SIFT_DetectAndCompute_Async(SIFT self, Mat src, Mat mask, CvCallback_2 callback) {
  BEGIN_WRAP
  std::vector<cv::KeyPoint> detected;
  cv::Mat desc;
  (*self.ptr)->detectAndCompute(*src.ptr, *mask.ptr, detected, desc);
  callback(veckeypoint_cpp2c_p(detected), new Mat{new cv::Mat(desc)});
  END_WRAP
}

File: lib/src/features2d/features2d_async.dart
// Copyright (c) 2024, rainyl and all contributors. All rights reserved.
// Use of this source code is governed by a Apache-2.0 license
// that can be found in the LICENSE file.

// ignore_for_file: non_constant_identifier_names

library cv.features2d;

import '../core/base.dart';
import '../core/dmatch.dart';
import '../core/keypoint.dart';
import '../core/mat.dart';
import '../core/scalar.dart';
import '../core/vec.dart';
import '../g/features2d.g.dart' as cfeatures2d;
import './features2d.dart';

extension AKAZEAsync on AKAZE {
  /// Detect keypoints in an image using AKAZE.
  ///
  /// For further details, please see:
  /// https://docs.opencv.org/master/d0/d13/classcv_1_1Feature2D.html#aa4e9a7082ec61ebc108806704fbd7887
  Future<VecKeyPoint> detectAsync(Mat src) async {
    final ret = VecKeyPoint();
    return cvRunAsync0<VecKeyPoint>(
      (callback) => cfeatures2d.cv_AKAZE_detect(ref, src.ref, ret.ptr, callback),
      (c) {
        return c.complete(ret);
      },
    );
  }

  /// DetectAndCompute keypoints and compute in an image using AKAZE.
  ///
  /// For further details, please see:
  /// https://docs.opencv.org/master/d0/d13/classcv_1_1Feature2D.html#a8be0d1c20b08eb867184b8d74c15a677
  Future<(VecKeyPoint, Mat)> detectAndComputeAsync(Mat src, Mat mask) async {
    final desc = Mat.empty();
    final ret = VecKeyPoint();
    return cvRunAsync0(
      (callback) =>
          cfeatures2d.cv_AKAZE_detectAndCompute(ref, src.ref, mask.ref, desc.ref, ret.ptr, callback),
      (c) {
        return c.complete((ret, desc));
      },
    );
  }
}

extension AgastFeatureDetectorAsync on AgastFeatureDetector {
  /// Detect keypoints in an image using AgastFeatureDetector.
  ///
  /// For further details, please see:
  /// https://docs.opencv.org/master/d0/d13/classcv_1_1Feature2D.html#aa4e9a7082ec61ebc108806704fbd7887
  Future<VecKeyPoint> detectAsync(Mat src) async {
    final ret = VecKeyPoint();
    return cvRunAsync0(
      (callback) => cfeatures2d.cv_AgastFeatureDetector_detect(ref, src.ref, ret.ptr, callback),
      (c) {
        return c.complete(ret);
      },
    );
  }
}

extension BRISKAsync on BRISK {
  /// Detect keypoints in an image using BRISK.
  ///
  /// For further details, please see:
  /// https://docs.opencv.org/master/d0/d13/classcv_1_1Feature2D.html#aa4e9a7082ec61ebc108806704fbd7887
  Future<VecKeyPoint> detectAsync(Mat src) async {
    final ret = VecKeyPoint();
    return cvRunAsync0((callback) => cfeatures2d.cv_BRISK_detect(ref, src.ref, ret.ptr, callback), (c) {
      return c.complete(ret);
    });
  }

  /// DetectAndCompute keypoints and compute in an image using BRISK.
  ///
  /// For further details, please see:
  /// https://docs.opencv.org/master/d0/d13/classcv_1_1Feature2D.html#a8be0d1c20b08eb867184b8d74c15a677
  Future<(VecKeyPoint, Mat)> detectAndComputeAsync(Mat src, Mat mask) async {
    final desc = Mat.empty();
    final ret = VecKeyPoint();
    return cvRunAsync0(
      (callback) =>
          cfeatures2d.cv_BRISK_detectAndCompute(ref, src.ref, mask.ref, desc.ref, ret.ptr, callback),
      (c) {
        return c.complete((ret, desc));
      },
    );
  }
}

extension FastFeatureDetectorAsync on FastFeatureDetector {
  /// Detect keypoints in an image using FastFeatureDetector.
  ///
  /// For further details, please see:
  /// https://docs.opencv.org/master/d0/d13/classcv_1_1Feature2D.html#aa4e9a7082ec61ebc108806704fbd7887
  Future<VecKeyPoint> detectAsync(Mat src) async {
    final ret = VecKeyPoint();
    return cvRunAsync0(
      (callback) => cfeatures2d.cv_FastFeatureDetector_detect(ref, src.ref, ret.ptr, callback),
      (c) {
        return c.complete(ret);
      },
    );
  }
}

extension GFTTDetectorAsync on GFTTDetector {
  /// Detect keypoints in an image using GFTTDetector.
  ///
  /// For further details, please see:
  /// https://docs.opencv.org/master/d0/d13/classcv_1_1Feature2D.html#aa4e9a7082ec61ebc108806704fbd7887
  Future<VecKeyPoint> detectAsync(Mat src) async {
    final ret = VecKeyPoint();
    return cvRunAsync0((callback) => cfeatures2d.cv_GFTTDetector_detect(ref, src.ref, ret.ptr, callback), (
      c,
    ) {
      return c.complete(ret);
    });
  }
}

extension KAZEAsync on KAZE {
  /// Detect keypoints in an image using KAZE.
  ///
  /// For further details, please see:
  /// https://docs.opencv.org/master/d0/d13/classcv_1_1Feature2D.html#aa4e9a7082ec61ebc108806704fbd7887
  Future<VecKeyPoint> detectAsync(Mat src) async {
    final ret = VecKeyPoint();
    return cvRunAsync0((callback) => cfeatures2d.cv_KAZE_detect(ref, src.ref, ret.ptr, callback), (c) {
      return c.complete(ret);
    });
  }

  /// DetectAndCompute keypoints and compute in an image using KAZE.
  ///
  /// For further details, please see:
  /// https://docs.opencv.org/master/d0/d13/classcv_1_1Feature2D.html#a8be0d1c20b08eb867184b8d74c15a677
  Future<(VecKeyPoint, Mat)> detectAndComputeAsync(Mat src, Mat mask) async {
    final desc = Mat.empty();
    final ret = VecKeyPoint();
    return cvRunAsync0(
      (callback) => cfeatures2d.cv_KAZE_detectAndCompute(ref, src.ref, mask.ref, desc.ref, ret.ptr, callback),
      (c) {
        return c.complete((ret, desc));
      },
    );
  }
}

extension MSERAsync on MSER {
  /// Detect keypoints in an image using MSER.
  ///
  /// For further details, please see:
  /// https://docs.opencv.org/master/d0/d13/classcv_1_1Feature2D.html#aa4e9a7082ec61ebc108806704fbd7887
  Future<VecKeyPoint> detectAsync(Mat src) async {
    final ret = VecKeyPoint();
    return cvRunAsync0((callback) => cfeatures2d.cv_MSER_detect(ref, src.ref, ret.ptr, callback), (c) {
      return c.complete(ret);
    });
  }
}

extension ORBAsync on ORB {
  /// Detect keypoints in an image using ORB.
  ///
  /// For further details, please see:
  /// https://docs.opencv.org/master/d0/d13/classcv_1_1Feature2D.html#aa4e9a7082ec61ebc108806704fbd7887
  Future<VecKeyPoint> detectAsync(Mat src) async {
    final ret = VecKeyPoint();
    return cvRunAsync0((callback) => cfeatures2d.cv_ORB_detect(ref, src.ref, ret.ptr, callback), (c) {
      return c.complete(ret);
    });
  }

  /// DetectAndCompute keypoints and compute in an image using ORB.
  ///
  /// For further details, please see:
  /// https://docs.opencv.org/master/d0/d13/classcv_1_1Feature2D.html#a8be0d1c20b08eb867184b8d74c15a677
  Future<(VecKeyPoint, Mat)> detectAndComputeAsync(
    Mat src,
    Mat mask, {
    VecKeyPoint? keypoints,
    Mat? description,
    bool useProvidedKeypoints = false,
  }) async {
    keypoints ??= VecKeyPoint();
    description ??= Mat.empty();
    return cvRunAsync0(
      (callback) => cfeatures2d.cv_ORB_detectAndCompute(
        ref,
        src.ref,
        mask.ref,
        keypoints!.ptr,
        description!.ref,
        useProvidedKeypoints,
        callback,
      ),
      (c) {
        return c.complete((keypoints!, description!));
      },
    );
  }
}

extension SimpleBlobDetectorAsync on SimpleBlobDetector {
  /// Detect keypoints in an image using SimpleBlobDetector.
  ///
  /// For further details, please see:
  /// https://docs.opencv.org/master/d0/d13/classcv_1_1Feature2D.html#aa4e9a7082ec61ebc108806704fbd7887
  Future<VecKeyPoint> detectAsync(Mat src) async {
    final ret = VecKeyPoint();
    return cvRunAsync0(
      (callback) => cfeatures2d.cv_SimpleBlobDetector_detect(ref, src.ref, ret.ptr, callback),
      (c) {
        return c.complete(ret);
      },
    );
  }
}

extension BFMatcherAsync on BFMatcher {
  /// Match Finds the best match for each descriptor from a query set.
  ///
  /// For further details, please see:
  /// https://docs.opencv.org/4.x/db/d39/classcv_1_1DescriptorMatcher.html#a0f046f47b68ec7074391e1e85c750cba
  Future<VecDMatch> matchAsync(Mat query, Mat train) async {
    final ret = VecDMatch();
    return cvRunAsync0(
      (callback) => cfeatures2d.cv_BFMatcher_match(ref, query.ref, train.ref, ret.ptr, callback),
      (c) {
        return c.complete(ret);
      },
    );
  }

  /// KnnMatch Finds the k best matches for each descriptor from a query set.
  ///
  /// For further details, please see:
  /// https://docs.opencv.org/master/db/d39/classcv_1_1DescriptorMatcher.html#aa880f9353cdf185ccf3013e08210483a
  Future<VecVecDMatch> knnMatchAsync(Mat query, Mat train, int k) async {
    final ret = VecVecDMatch();
    return cvRunAsync0(
      (callback) => cfeatures2d.cv_BFMatcher_knnMatch(ref, query.ref, train.ref, k, ret.ptr, callback),
      (c) {
        return c.complete(ret);
      },
    );
  }
}

extension FlannBasedMatcherAsync on FlannBasedMatcher {
  /// KnnMatch Finds the k best matches for each descriptor from a query set.
  ///
  /// For further details, please see:
  /// https://docs.opencv.org/master/db/d39/classcv_1_1DescriptorMatcher.html#aa880f9353cdf185ccf3013e08210483a
  Future<VecVecDMatch> knnMatchAsync(Mat query, Mat train, int k) async {
    final ret = VecVecDMatch();
    return cvRunAsync0(
      (callback) =>
          cfeatures2d.cv_FlannBasedMatcher_knnMatch(ref, query.ref, train.ref, k, ret.ptr, callback),
      (c) {
        return c.complete(ret);
      },
    );
  }
}

extension SIFTAsync on SIFT {
  /// Detect keypoints in an image using SIFT.
  ///
  /// For further details, please see:
  /// https://docs.opencv.org/master/d0/d13/classcv_1_1Feature2D.html#aa4e9a7082ec61ebc108806704fbd7887
  Future<VecKeyPoint> detectAsync(Mat src) async {
    final ret = VecKeyPoint();
    return cvRunAsync0((callback) => cfeatures2d.cv_SIFT_detect(ref, src.ref, ret.ptr, callback), (c) {
      return c.complete(ret);
    });
  }

  /// DetectAndCompute keypoints and compute in an image using SIFT.
  ///
  /// For further details, please see:
  /// https://docs.opencv.org/master/d0/d13/classcv_1_1Feature2D.html#a8be0d1c20b08eb867184b8d74c15a677
  Future<(VecKeyPoint, Mat)> detectAndComputeAsync(Mat src, Mat mask) async {
    final desc = Mat.empty();
    final ret = VecKeyPoint();
    return cvRunAsync0(
      (callback) => cfeatures2d.cv_SIFT_detectAndCompute(ref, src.ref, mask.ref, desc.ref, ret.ptr, callback),
      (c) {
        return c.complete((ret, desc));
      },
    );
  }
}

Future<void> drawKeyPointsAsync(
  Mat src,
  VecKeyPoint keypoints,
  Mat dst,
  Scalar color,
  DrawMatchesFlag flag,
) async {
  await cvRunAsync0<void>(
    (callback) =>
        cfeatures2d.cv_drawKeyPoints(src.ref, keypoints.ref, dst.ref, color.ref, flag.value, callback),
    (c) => c.complete(),
  );
}

/// DrawMatches draws matches on combined train and querry images.
///
/// For further details, please see:
/// https://docs.opencv.org/master/d4/d5d/group__features2d__draw.html#gad8f463ccaf0dc6f61083abd8717c261a
Future<void> drawMatchesAsync(
  InputArray img1,
  VecKeyPoint keypoints1,
  InputArray img2,
  VecKeyPoint keypoints2,
  VecDMatch matches1to2,
  InputOutputArray outImg, {
  Scalar? matchColor,
  Scalar? singlePointColor,
  VecChar? matchesMask,
  DrawMatchesFlag flags = DrawMatchesFlag.DEFAULT,
}) async {
  matchColor ??= Scalar.all(-1);
  singlePointColor ??= Scalar.all(-1);
  matchesMask ??= VecChar.fromList([]);
  return cvRunAsync0(
    (callback) => cfeatures2d.cv_drawMatches(
      img1.ref,
      keypoints1.ref,
      img2.ref,
      keypoints2.ref,
      matches1to2.ref,
      outImg.ref,
      matchColor!.ref,
      singlePointColor!.ref,
      matchesMask!.ref,
      flags.value,
      callback,
    ),
    (c) {
      c.complete();
    },
  );
}
